### ISSUE TITLE
General feedback

### ISSUE THREAD

**Issue by hanbzu** [anshulsharma-kantar/premier-league#1]
Hello! I'm posting some feedback on the code here. I hope it's helpful to you :-)

**Comment by hanbzu** [comment]
### Minimise state

We need to make the application feel snappy, quick to react to user actions. To do so, when we create components we should try to minimise state transitions, because they slow down the app.

And the best strategy towards minimising state transitions is **minimising state.**

For example, in `Teampage` component, you have 6 bits of state:

https://github.com/anshulsharma-kantar/premier-league/blob/560ccbda5a00adf957588d57e519b280cbab90c0/src/screens/team-page.js#L29-L34

Each of these, when it updates will make the whole component re-render.

We can refactor it and remove 2 unnecessary states:

```jsx
const { data, error, isLoading } = useSWR("teams", getTeamData);
// For the detailed view, we only need to hold the index or null (=details not shown)
const [detailsFor, setDetailsFor] = React.useState(null); // Detailed view

// We keep these visualisation settings needed for the table
// (but it'd be better to extract them to a PaginatedTable component)
const [page, setPage] = React.useState(0);
const [rowsPerPage, setRowsPerPage] = React.useState(10);

// Instead of adding state, I can get a sorted version of the data in another variable
const rows = data && [...data].sort((a, b) => a.position - b.position)
```
 And here's how you'd render the `TeamDetails` part given that `detailsFor` only holds an index:

```jsx
<TeamDetails
  isOpen={Boolean(detailsFor)}
  handleModalClose={() => setDetailsFor(null)}
  teamData={Boolean(detailsFor) && rows?.[detailsFor]}
/>
```

**TL;DR: Try to find the minimum state that represents your interactions, try to store string/numerical/boolean state if you can over arrays or objects (because they're cheaper to compare). For the rest of data transformations, derive views from that minimal state, using immutable operations of the state.**

**Comment by hanbzu** [comment]
### Preserve the context by giving good names

Or, alternativelly **use inline code.** Example:

https://github.com/anshulsharma-kantar/premier-league/blob/560ccbda5a00adf957588d57e519b280cbab90c0/src/screens/team-page.js#L61

There are many _clicks_ on this component. You could fix this by either calling this function something like `handleClickOnRow` or (my preference) by **inlining the code** so it's clear where the click comes from:

```jsx
<TableRow
  hover
  tabIndex={-1}
  onClick={() => setDetailsFor(i) /* ← This way I see what clicking on TableRow does */ } 
  key={row.id}
>
```

**Comment by hanbzu** [comment]
### For async state, use `useSWR` where you can

We're already using this library for data fetching anyway and it's a well-maintained library for memoised async state.

For example, although `useAsync` may be a good way of handling async state, `useSWR` is already giving us this, with fewer lines of code and handling much of the complexity in `auth-provider.js`:
 
https://github.com/anshulsharma-kantar/premier-league/blob/560ccbda5a00adf957588d57e519b280cbab90c0/src/context/auth-context.js#L40

It was probably a good exercise to build these hooks yourself. Still, we're going to try to use `useSWR` moving forward for these cases, also because these things are evolving within React and `useSWR` is going to provide us a more stable API as the react team improves async hooks.

Also, here's another example of where you should have used `useSWR` (giving you instant cached data):

https://github.com/anshulsharma-kantar/premier-league/blob/560ccbda5a00adf957588d57e519b280cbab90c0/src/screens/team-depth.js#L18-L26

Like this:

```jsx
 const {data: teamData} = useSWR(...)
```

**Comment by hanbzu** [comment]
### Code comments and docs

I value written documentation a lot. And if it's located close to the code (as in `.md` files close to the concern areas) or in the code as comments it's even quite robust to changes.

Yes, probably good code should have great names and should be pretty understandable, but sometimes it's cool to leave out some notes of what your intention was to your colleagues or reviewers, especially in more obscure areas.

Also, `README.md` still shows the boilerplate docs, it'd be nice to see how to run the project there instead and what external tools you need to run and in what port/order, etc.

**Comment by hanbzu** [comment]
### Remember to clean up old code and warnings

These leftovers make it hard to read the code and maintain a tidy repo.

<img width="523" alt="image" src="https://user-images.githubusercontent.com/684539/220101095-9c6c8b9b-fcc9-4c74-b76e-fb125d774b88.png">

**Comment by hanbzu** [comment]
### Extract patterns to their own component

You should be making the call of when this does or does not sense, but the focus should be **readability** and **maintainability**, or said otherwise, _think on your colleagues and the future_ :-)

For example, here, this code is repeated many times:

https://github.com/anshulsharma-kantar/premier-league/blob/560ccbda5a00adf957588d57e519b280cbab90c0/src/screens/team-depth.js#L44-L51
